### PR TITLE
Unit tests for database shards

### PR DESCRIPTION
**Pull request by p2peer**

### CODE REVIEW

**Comment by miguelportilla** [comment]
This PR has a dependency on PR 3252

**Comment by carlhua** [comment]
Travis CI now passing with updated time out value.

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp [hunk 1]
```diff
@@ -143,14 +143,21 @@ DatabaseShardImp::init()
...
-                        return false;
-
-                    // Remove legacy shard
                     shard->removeOnDestroy();
-                    JLOG(j_.warn()) <<
-                        "shard " << shardIndex <<
-                        " removed, legacy shard";
+                    if (!shard->isLegacy())
+                    {
+                        // Remove corrupted shard
```
Comment (undertome) [inline src/ripple/nodestore/impl/DatabaseShardImp.cpp:149]: These comments (lines 149 and 156) are somewhat misleading as the removal isn't happening here.

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp [hunk 2]
```diff
@@ -143,14 +143,21 @@ DatabaseShardImp::init()
...
+                        JLOG(j_.warn()) <<
+                            "shard " << shardIndex <<
+                            " removed, corrupted shard";
+                    }
+                    else
+                    {
+                        // Remove legacy shard
+                        JLOG(j_.warn()) <<
```
Comment (undertome) [inline src/ripple/nodestore/impl/DatabaseShardImp.cpp:157]: Could you refactor these two log messages into a single statement? You can use a ternary operator to pick whether to output "corrupted" or "legacy".

File: src/test/nodestore/DatabaseShard_test.cpp [hunk 3]
```diff
@@ -0,0 +1,894 @@
...
+#include <ripple/core/ConfigSections.h>
+
+namespace ripple {
+namespace NodeStore {
+
+// Tests DatabaseShard class
+//
+class DatabaseShard_test : public TestBase
+{
+    static constexpr std::uint32_t max_size_gb = 10;
```
Comment (undertome) [inline src/test/nodestore/DatabaseShard_test.cpp:40]: camelCase rather than snake_case seems to be the repo convention.

File: src/test/nodestore/DatabaseShard_test.cpp [hunk 4]
```diff
@@ -0,0 +1,894 @@
...
+            for (std::uint32_t i = 0; i < 4 * ledgers_per_shard; ++i)
+                check_ledger(D, db, D.ledgers_[i]);
+        }
+    }
+
+    void test_all (std::string const& backendType)
+    {
+        std::uint64_t const seedValue = 51;
+        test_standalone(backendType);
+        test_create_shard(backendType, seedValue);
```
Comment (undertome) [inline src/test/nodestore/DatabaseShard_test.cpp:862]: There are some inconsistencies in the repo but the majority of method names seem to be camelCase as well.

File: Builds/CMake/RippledCore.cmake [hunk 5]
```diff
@@ -844,6 +844,7 @@ target_sources (rippled PRIVATE
   src/test/nodestore/Backend_test.cpp
   src/test/nodestore/Basics_test.cpp
   src/test/nodestore/Database_test.cpp
+  src/test/nodestore/DatabaseShard_test.cpp
```
Comment (undertome) [inline Builds/CMake/RippledCore.cmake:847]: Nit: this should go before Database_test.cpp

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp [hunk 6]
```diff
@@ -1296,6 +1306,9 @@ DatabaseShardImp::setFileStats()
         for (auto const& e : shards_)
             if (e.second.shard)
                 shards.push_back(e.second.shard);
+
+        if (shards.empty())
```
Comment (undertome) [inline src/ripple/nodestore/impl/DatabaseShardImp.cpp:1310]: Why are you duplicating this check which occurs also at line 1303?

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp [hunk 7]
```diff
@@ -1109,6 +1116,9 @@ DatabaseShardImp::initConfig(std::lock_guard<std::mutex>&)
         ledgersPerShard_ = get<std::uint32_t>(section, "ledgers_per_shard");
         if (ledgersPerShard_ == 0 || ledgersPerShard_ % 256 != 0)
             return fail("'ledgers_per_shard' must be a multiple of 256");
+
+        earliestShardIndex_ = seqToShardIndex(earliestLedgerSeq());
```
Comment (undertome) [inline src/ripple/nodestore/impl/DatabaseShardImp.cpp:1120]: Why are you overwriting `earliestShardIndex_` which should already be initialized in the constructor?

File: src/ripple/nodestore/impl/DatabaseShardImp.h [hunk 8]
```diff
@@ -232,7 +232,7 @@ class DatabaseShardImp : public DatabaseShard
     std::uint32_t ledgersPerShard_ = ledgersPerShardDefault;
 
     // The earliest shard index
-    std::uint32_t const earliestShardIndex_;
+    std::uint32_t earliestShardIndex_;
```
Comment (undertome) [inline src/ripple/nodestore/impl/DatabaseShardImp.h:235]: It looks like you removed the `const` qualifier in order to set `earliestShardIndex_` in `DatabaseShardImp::initConfig`. Why is that necessary when this variable gets initialized during the constructor?

File: src/test/jtx/impl/Env.cpp [hunk 9]
```diff
@@ -124,11 +125,17 @@ Env::close(NetClock::time_point closeTime,
...
     else
     {
-        rpc("ledger_accept");
+        auto resp = rpc("ledger_accept");
+        if (resp["result"]["status"] != std::string("success"))
+        {
+            std::cerr << "Env::close() failed: " << resp["result"]["status"] << std::endl;
+            res = false;
+        }
         // VFALCO No error check?
```
Comment (undertome) [inline src/test/jtx/impl/Env.cpp:134]: Looks like this comment can be removed now

File: src/test/nodestore/DatabaseShard_test.cpp [hunk 10]
```diff
@@ -0,0 +1,894 @@
...
+            {
+                env_(pay(*A_[pay_[seq][i].first], *A_[pay_[seq][i].second], XRP(xrp_[seq])));
+            }
+        }
+
+        bool make_ledgers(test::jtx::Env &env_)
+        {
+            for (std::uint32_t i = 3; i <= ledgers_per_shard; ++i)
+            {
+                env_.close();
```
Comment (undertome) [inline src/test/nodestore/DatabaseShard_test.cpp:129]: You modified the `close` method to return a `bool` indicating success or failure. Might as well confirm the return value.

File: src/test/nodestore/DatabaseShard_test.cpp [hunk 11]
```diff
@@ -0,0 +1,894 @@
...
+        void make_ledger_data(test::jtx::Env &env_, std::uint32_t seq)
+        {
+            using namespace test::jtx;
+
+            if (new_acc(seq))
+                 env_.fund(XRP(ini_amount), *A_[acc_[seq] - 1]);
+
+            for (std::uint32_t i = 0; i < pay_[seq].size(); ++i)
+            {
+                env_(pay(*A_[pay_[seq][i].first], *A_[pay_[seq][i].second], XRP(xrp_[seq])));
```
Comment (undertome) [inline src/test/nodestore/DatabaseShard_test.cpp:121]: Please run this entire file under clang format; the 80 character width limit and other conventions are not being followed.

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp (hunk 7 above)
Comment (p2peer) [reply src/ripple/nodestore/impl/DatabaseShardImp.cpp:1120]: Value of earliestShardIndex_ depends on parameter ledgers_per_shard which is defined to default value of 16384 in constructor. But for test we redefine ledgers_per_shard to 256. This redefinition applies to standalone mode only, so it is done not in constructor, but in init() function. After redefinition of  ledgers_per_shard, the value of earliestShardIndex_ should also be rewritten.

File: src/ripple/nodestore/impl/DatabaseShardImp.h (hunk 8 above)
Comment (p2peer) [reply src/ripple/nodestore/impl/DatabaseShardImp.h:235]: See previous comment.

File: src/test/nodestore/DatabaseShard_test.cpp (hunk 11 above)
Comment (p2peer) [reply src/test/nodestore/DatabaseShard_test.cpp:121]: Done

File: src/test/nodestore/DatabaseShard_test.cpp (hunk 10 above)
Comment (p2peer) [reply src/test/nodestore/DatabaseShard_test.cpp:129]: Done

File: src/test/jtx/impl/Env.cpp (hunk 9 above)
Comment (p2peer) [reply src/test/jtx/impl/Env.cpp:134]: Done

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp (hunk 6 above)
Comment (p2peer) [reply src/ripple/nodestore/impl/DatabaseShardImp.cpp:1310]: Done

File: src/test/nodestore/DatabaseShard_test.cpp (hunk 4 above)
Comment (p2peer) [reply src/test/nodestore/DatabaseShard_test.cpp:862]: Done

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp (hunk 2 above)
Comment (p2peer) [reply src/ripple/nodestore/impl/DatabaseShardImp.cpp:157]: Done

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp (hunk 1 above)
Comment (p2peer) [reply src/ripple/nodestore/impl/DatabaseShardImp.cpp:149]: Done

**Comment by p2peer** [comment]
> Looks good to me.
> 
> The new test is now our longest running test by quite a margin. If there is any way to make the test run faster, I'd be in favor of that. However I wouldn't want to do that if it meant compromising the quality of the test. I took a look, and didn't see any obvious way to make the test run faster.

I reduced test time up to 5 times.

**Comment by p2peer** [comment]
> Travis CI now passing with updated time out value.

Thanks

File: Builds/CMake/RippledCore.cmake (hunk 5 above)
Comment (p2peer) [reply Builds/CMake/RippledCore.cmake:847]: Done

File: src/test/nodestore/DatabaseShard_test.cpp [hunk 12]
```diff
@@ -0,0 +1,986 @@
...
+//
+class DatabaseShard_test : public TestBase
+{
+    static constexpr std::uint32_t maxSizeGb = 10;
+    static constexpr std::uint32_t ledgersPerShard = 256;
+    static constexpr std::uint32_t earliestSeq = ledgersPerShard + 1;
+    static constexpr std::uint32_t dataSizeMax = 4;
+    static constexpr std::uint32_t iniAmount = 1000000;
+    static constexpr std::uint32_t nTestShards = 4;
+    static constexpr std::uint32_t shardStoreTimeout = 60;
```
Comment (HowardHinnant) [inline src/test/nodestore/DatabaseShard_test.cpp:46]: Consider using the std::chrono timing API: https://github.com/HowardHinnant/rippled/commit/d43f1682065b8ace5d5d8360ed6609b4e69b7682

File: .travis.yml [hunk 13]
```diff
@@ -408,7 +408,7 @@ cache:
     - $CACHE_DIR
 
 before_install:
-  - if [ "$(uname)" = "Darwin" ] ; then export NUM_PROCESSORS=$(sysctl -n hw.physicalcpu); else export NUM_PROCESSORS=$(nproc); fi
+  - if [ "$(uname)" = "Darwin" ] ; then export NUM_PROCESSORS=1; else export NUM_PROCESSORS=$(nproc); fi
```
Comment (seelabs) [inline .travis.yml:411]: What's the motivation for this change? (apologies if this has been discussed, just refer me to a previous discussion if so). At any rate, a comment explaining the change would be good.

File: src/ripple/nodestore/impl/DatabaseShardImp.h (hunk 8 above)
Comment (seelabs) [reply src/ripple/nodestore/impl/DatabaseShardImp.h:235]: I understand why you did this - it's a consequence of the two phase initialization and testing. I'm OK with the change, but I'd add a comment explaining why this is not const.

File: src/test/jtx/Env.h [hunk 14]
```diff
@@ -326,7 +326,7 @@ class Env
             The Application network time is set to
             the close time of the resulting ledger.
     */
-    void
+    bool
```
Comment (seelabs) [inline src/test/jtx/Env.h:329]: Needs a brief comment explaining what the return value does (true, no error; false error).

File: src/test/nodestore/DatabaseShard_test.cpp [hunk 15]
```diff
@@ -0,0 +1,986 @@
...
+        BEAST_EXPECT(setCount == newacc);
+        BEAST_EXPECT(payCount == D.pay_[seq].size());
+        BEAST_EXPECT(tothCount == !seq);
+    }
+
+    bool
+    saveLedger(
+        Database* db,
+        std::shared_ptr<const Ledger> l,
+        std::shared_ptr<const Ledger> next = {})
```
Comment (seelabs) [inline src/test/nodestore/DatabaseShard_test.cpp:241]: Couple nits:
1) `db` by reference
2) pass the two  `shared_ptr` by `const&` (passing by value requires incrementing atomic variables)
3) Do `l` and `next` need to be `shared_ptr`? Could be just pass them as `Ledger const&`?
4) `const` goes on the right. I.e. `Ledger const` not `const Ledger`

File: src/test/nodestore/DatabaseShard_test.cpp [hunk 16]
```diff
@@ -0,0 +1,986 @@
...
+        auto accCount{0};
+        auto sothCount{0};
+        for (auto const& s : l->sles)
+        {
+            if (s->getType() == ltACCOUNT_ROOT)
+            {
+                int sq = s->getFieldU32(sfSequence);
+                int reqsq = -1;
+                const auto id = s->getAccountID(sfAccount);
+                int i;
```
Comment (seelabs) [inline src/test/nodestore/DatabaseShard_test.cpp:173]: I'd put this declaration in the loop. I.e. `for (int i=0;...`

File: src/test/nodestore/DatabaseShard_test.cpp [hunk 17]
```diff
@@ -0,0 +1,986 @@
...
+            if (!l->txMap().isValid())
+                return false;
+            l->txMap().snapShot(false)->visitNodes(visitTx);
+        }
+
+        return true;
+    }
+
+    void
+    checkLedger(testData& D, DatabaseShard* db, std::shared_ptr<const Ledger> l)
```
Comment (seelabs) [inline src/test/nodestore/DatabaseShard_test.cpp:308]: See comment on `saveLedger`

File: src/test/nodestore/DatabaseShard_test.cpp [hunk 18]
```diff
@@ -0,0 +1,986 @@
...
+    static constexpr std::uint32_t ledgersPerShard = 256;
+    static constexpr std::uint32_t earliestSeq = ledgersPerShard + 1;
+    static constexpr std::uint32_t dataSizeMax = 4;
+    static constexpr std::uint32_t iniAmount = 1000000;
+    static constexpr std::uint32_t nTestShards = 4;
+    static constexpr std::uint32_t shardStoreTimeout = 60;
+    test::SuiteJournal journal_;
+    beast::temp_dir defNodeDir;
+
+    struct testData
```
Comment (seelabs) [inline src/test/nodestore/DatabaseShard_test.cpp:50]: `testData` -> `TestData`

File: src/test/nodestore/DatabaseShard_test.cpp [hunk 19]
```diff
@@ -0,0 +1,986 @@
...
+            if (nodeDir.empty())
+                cfg->overwrite(
+                    ConfigSection::nodeDatabase(), "path", defNodeDir.path());
+            else
+                cfg->overwrite(ConfigSection::nodeDatabase(), "path", nodeDir);
+            return cfg;
+        });
+    }
+
+    int
```
Comment (seelabs) [inline src/test/nodestore/DatabaseShard_test.cpp:450]: I prefer `optional` to returning `-1`. Fine for this test code if you want to keep it tho.

File: src/test/nodestore/DatabaseShard_test.cpp [hunk 20]
```diff
@@ -0,0 +1,986 @@
...
+        {
+            if (!BEAST_EXPECT(time(0) - time0 < timeout))
+                return -1;
+            std::this_thread::yield();
+        }
+
+        return shardNumber;
+    }
+
+    int
```
Comment (seelabs) [inline src/test/nodestore/DatabaseShard_test.cpp:469]: `optional`

File: src/test/nodestore/DatabaseShard_test.cpp (hunk 12 above)
Comment (p2peer) [reply src/test/nodestore/DatabaseShard_test.cpp:46]: Done

File: .travis.yml (hunk 13 above)
Comment (p2peer) [reply .travis.yml:411]: This change was made due to problems of launching unit tests in parallel on Mac platform, see the thread https://ripplelabs.slack.com/archives/C02GE9NBA/p1586457327294700

Done comment on .travis.yml.

File: src/test/jtx/Env.h (hunk 14 above)
Comment (p2peer) [reply src/test/jtx/Env.h:329]: Done

File: src/test/nodestore/DatabaseShard_test.cpp [hunk 21]
```diff
@@ -0,0 +1,986 @@
...
+    static constexpr std::uint32_t nTestShards = 4;
+    static constexpr std::uint32_t shardStoreTimeout = 60;
+    test::SuiteJournal journal_;
+    beast::temp_dir defNodeDir;
+
+    struct testData
+    {
+        beast::xor_shift_engine rng_;
+        int nShards_;
+        std::vector<std::shared_ptr<test::jtx::Account>> A_;
```
[the remark this answers is not in the document]
Comment (p2peer) [reply src/test/nodestore/DatabaseShard_test.cpp:54]: Done

File: src/test/nodestore/DatabaseShard_test.cpp [hunk 22]
```diff
@@ -0,0 +1,986 @@
...
+        std::vector<int> acc_;
+        std::vector<std::vector<std::pair<int, int>>> pay_;
```
[the remark this answers is not in the document]
Comment (p2peer) [reply src/test/nodestore/DatabaseShard_test.cpp:56]: Done

File: src/test/nodestore/DatabaseShard_test.cpp [hunk 23]
```diff
@@ -0,0 +1,986 @@
...
+                {
+                    char str[9];
+                    for (int j = 0; j < 8; ++j)
+                        str[j] = 'a' + rand_int(rng_, 'z' - 'a');
+                    str[8] = 0;
+                    A_.push_back(std::make_shared<test::jtx::Account>(str));
+                }
+
+                acc_.push_back(n);
+                pay_.push_back(pay);
```
[the remark this answers is not in the document]
Comment (p2peer) [reply src/test/nodestore/DatabaseShard_test.cpp:101]: Done

File: src/test/nodestore/DatabaseShard_test.cpp [hunk 24]
```diff
@@ -0,0 +1,986 @@
...
+                    return false;
+                ledgers_.push_back(ledger);
+            }
+
+            return true;
+        }
+    };
+
+    void
+    testLedgerData(testData& D, std::shared_ptr<Ledger> l, std::uint32_t seq)
```
[the remark this answers is not in the document]
Comment (p2peer) [reply src/test/nodestore/DatabaseShard_test.cpp:159]: Done

File: src/test/nodestore/DatabaseShard_test.cpp (hunk 16 above)
Comment (p2peer) [reply src/test/nodestore/DatabaseShard_test.cpp:173]: Done

File: src/test/nodestore/DatabaseShard_test.cpp (hunk 18 above)
Comment (p2peer) [reply src/test/nodestore/DatabaseShard_test.cpp:50]: Done

File: src/test/nodestore/DatabaseShard_test.cpp (hunk 15 above)
Comment (p2peer) [reply src/test/nodestore/DatabaseShard_test.cpp:241]: Done

File: src/test/nodestore/DatabaseShard_test.cpp (hunk 17 above)
Comment (p2peer) [reply src/test/nodestore/DatabaseShard_test.cpp:308]: Done

File: src/test/nodestore/DatabaseShard_test.cpp (hunk 19 above)
Comment (p2peer) [reply src/test/nodestore/DatabaseShard_test.cpp:450]: Done

File: src/test/nodestore/DatabaseShard_test.cpp (hunk 20 above)
Comment (p2peer) [reply src/test/nodestore/DatabaseShard_test.cpp:469]: Done

File: src/test/nodestore/DatabaseShard_test.cpp [hunk 25]
```diff
@@ -0,0 +1,986 @@
...
+            std::uint32_t n = 0;
+            for (std::uint32_t i = 0; i < ledgersPerShard * nShards; ++i)
+            {
+                int p;
+                if (n >= 2)
+                    p = rand_int(rng_, 2 * dataSize);
+                else
+                    p = 0;
+
+                std::vector<std::pair<int, int>> pay;
```
[the remark this answers is not in the document]
Comment (p2peer) [reply src/test/nodestore/DatabaseShard_test.cpp:75]: Done

**Comment by p2peer** [comment]
Depends on #3415 .

**Comment by miguelportilla** [comment]
This PR depends on PR 3415 being merged first.
